src/Workspaces/Core/Portable/Shared/Extensions/SyntaxEditorExtensions.cs

```diff
@@ -13,6 +13,29 @@ namespace Microsoft.CodeAnalysis.Shared.Extensions
 {
     internal static class SyntaxEditorExtensions
     {
+        /// <summary>
+        /// Performs several edits to a document.  If multiple edits are made within the same
+        /// expression context, then the document/semantic-model will be forked after each edit 
+        /// so that further edits can see if they're still safe to apply.
+        /// </summary>
+        public static Task ApplyExpressionLevelSemanticEditsAsync<TType>(
+            this SyntaxEditor editor, Document document,
+            ImmutableArray<TType> originalNodes,
+            Func<SemanticModel, TType, SyntaxNode, bool> canReplace,
+            Func<SemanticModel, SyntaxNode, TType, SyntaxNode, SyntaxNode> updateRoot,
+            Func<TType, SyntaxNode> selector,
```

Comment:
IMO, hte selector should go right after the 'originalNodes' parameter.  It's effectively one of the most important parts about this.  It is *the* thing that associated with originalNodes and maps accordingly.  Also, if it follows right afterwards, it can likely be generic on a TNode type which will then work properly with type inference (i think).

Finally, consider the following (but you don't have to do this if you don't want to): If i have a list of things and mapping from those things to another value, then that's just another way of saying "i have a mapping".  i.e. this may be more accurately expressed as a `ImmutableDictionary<TType, TNode>`.  Then, the simple case of just having array of nodes, is just a mapping from node to itself.  

That may simplify things by removing the need for multiple arguments by consolidating this into a single clearer argument.